openthread-border-router: init at unstable-2024-10-18

mrene opened the pull request:
## Description of changes
This adds [OpenThread](https://openthread.io/)'s open source [border router component](https://github.com/openthread/ot-br-posix) along with a NixOS service. 

Thread is a mesh networking technology based on 6LoWPAN (ipv6-based) and 802.15.4. The border router component bridges the mesh network to an ipv6 network on a local interface. If no ipv6 connectivity exists, it'll advertise a new prefix for the local network so communication is possible.

Once setup, it can be used to bridge a [Thread network](https://en.wikipedia.org/wiki/Thread_(network_protocol)) with the local network to expose IoT devices to other systems like Home Assistant, Apple Home, etc..

I've been using this implementation along with Home Assistant and some Nanoleaf light bulbs using both Matter and their proprietary "ltpdu" protocol for about a week now. It's pretty stable.

For the SkyConnect specifically, some options have to be used (see example). There doesn't seem to be a lot of hardware-specific options in nixpkgs so I chose not to include an option for this. Do let me know if there is an appropriate pattern to use here.

```nix
 services.openthread-border-router = {
   enable = true;
   radio =  {
     device = "/dev/serial/by-id/.....";
     baudRate = 460800;
     flowControl = true;
   };
 };
```

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

mrene committed:
openthread-border-router: init at unstable-2024-07-26

Pandapip1 reviewed (changes requested):
Move the changelog addition to its own commit

mrene committed:
rl-2605: mention openthread-border-router

mrene committed:
openthread-border-router: add nixos module

This can be used to use compatible Thread radios (such as HomeAssistant's SkyConnect)
in order to create a thread border router.

mrene commented:
I pushed a few changes:
- Added the packaged otbr-firewall script (with patches) in order to setup ip6tables rules referencing the otbr ipset
- Added the included otbr-web service which provides a web interface
- Cleaned up some of the cmake options which were already set by default 
- Split the changelog entry in its own commit
- Rebased over master

mrene commented on pkgs/by-name/op/openthread-border-router/package.nix:
```diff
@@ -0,0 +1,129 @@
+    (lib.cmakeFeature "CMAKE_CXX_STANDARD" "17")
+  ];
+
+  meta = with lib; {
+    description = "A Thread border router for POSIX-based platforms";
+    homepage = "https://github.com/openthread/ot-br-posix";
+    license = licenses.bsd3;
+    maintainers = with maintainers; [ mrene ];
+    mainProgram = "ot-ctl";
+    platforms = platforms.linux;
+  };
```
I couldn't get it to compile on darwin (some bad linker flag error but I didn't investigate further). I have no idea about BSDs, all the examples in the repository are linux-based distributions.

SuperSandro2000 commented on pkgs/by-name/op/openthread-border-router/package.nix:
```diff
@@ -0,0 +1,129 @@
+}:
+let
+  pname = "ot-br-posix";
+  version = "unstable-2024-10-18";
```
```suggestion
  version = "0-unstable-2024-10-18";
```

SuperSandro2000 commented on pkgs/by-name/op/openthread-border-router/package.nix:
```diff
@@ -0,0 +1,129 @@
+  };
+in
+stdenv.mkDerivation {
+
```
```suggestion
```

SuperSandro2000 commented on pkgs/by-name/op/openthread-border-router/package.nix:
```diff
@@ -0,0 +1,129 @@
+  env.NIX_CFLAGS_COMPILE = "-O";
+
+  patches = [
+    ./dont-install-systemd-units.patch
```
Why not?

mrene replied:
It would try to install them in their final location which fails in the sandbox

Pandapip1 replied:
Couldn't you modify the location to be inside the derivation?

mrene replied:
It would try to install them in their final location (not respecting the cmake install path), which fails in the sandbox

mrene replied:
> Couldn't you modify the location to be inside the derivation?

What for though? The units themselves are part of the nixos module.

Pandapip1 replied:
> What for though? The units themselves are part of the nixos module.

I could have sworn there was a standard path for systemd units. Oh well, feel free to resolve this.

SuperSandro2000 commented on pkgs/by-name/op/openthread-border-router/package.nix:
```diff
@@ -0,0 +1,129 @@
+  ];
+
+  meta = {
+    description = "A Thread border router for POSIX-based platforms";
```
```suggestion
    description = "Thread border router for POSIX-based platforms";
```

SuperSandro2000 commented on pkgs/by-name/op/openthread-border-router/package.nix:
```diff
@@ -0,0 +1,129 @@
+
+  postInstall = ''
+    mkdir -p $out/bin
+    cp ${src}/script/otbr-firewall $out/bin/
```
```suggestion
    cp ./script/otbr-firewall $out/bin/
```

SuperSandro2000 commented on pkgs/by-name/op/openthread-border-router/package.nix:
```diff
@@ -0,0 +1,129 @@
+    chmod +x $out/bin/otbr-firewall
+
+    # Patch the firewall script so we can run it within the systemd start script
+    sed -i 's/THREAD_IF=.*//' $out/bin/otbr-firewall
+    sed -i 's/.*init-functions//' $out/bin/otbr-firewall
+    sed -i 's/.*vars\.sh//' $out/bin/otbr-firewall
```
Can we use substituteInPlace or patches instead to notice if those break on changes? Also the patching should be done in postPatch.

SuperSandro2000 commented on pkgs/by-name/op/openthread-border-router/package.nix:
```diff
@@ -0,0 +1,129 @@
+
+  postPatch = ''
+    pushd third_party/openthread/repo
+      patch -p1 -i "${hass-addons}/openthread_border_router/0001-channel-monitor-disable-by-default.patch"
```
```suggestion
    patch -p1 -i "${hass-addons}/openthread_border_router/0001-channel-monitor-disable-by-default.patch"
```

SuperSandro2000 commented on nixos/modules/services/home-automation/openthread-border-router.nix:
```diff
@@ -0,0 +1,216 @@
+                "-d"
+                "${toString cfg.logLevel}"
+              ]
+              ++ (lib.optional (cfg.web.listenAddress != "") "-a ${cfg.web.listenAddress}")
+              ++ (lib.optional (cfg.web.listenPort != 0) "-p ${toString cfg.web.listenPort}")
```
```suggestion
              ++ lib.optional (cfg.web.listenAddress != "") "-a ${cfg.web.listenAddress}"
              ++ lib.optional (cfg.web.listenPort != 0) "-p ${toString cfg.web.listenPort}"
```

SuperSandro2000 commented on nixos/modules/services/home-automation/openthread-border-router.nix:
```diff
@@ -0,0 +1,216 @@
+          ExecStart = (
+            lib.concatStringsSep " " (
+              [ (lib.getExe' cfg.package "otbr-agent") ]
+              ++ [ "--verbose" ]
+              ++ [ "--backbone-ifname ${cfg.backboneInterface}" ]
+              ++ [ "--thread-ifname ${cfg.interfaceName}" ]
+              ++ [ "--debug-level ${toString cfg.logLevel}" ]
+              ++ (lib.optional (cfg.rest.listenPort != 0) "--rest-listen-port ${toString cfg.rest.listenPort}")
+              ++ (lib.optional (cfg.rest.listenAddress != "") "--rest-listen-address ${cfg.rest.listenAddress}")
```
```suggestion
              ++ [
                "--verbose"
                "--backbone-ifname ${cfg.backboneInterface}"
                "--thread-ifname ${cfg.interfaceName}"
                "--debug-level ${toString cfg.logLevel}"
              ]
              ++ lib.optional (cfg.rest.listenPort != 0) "--rest-listen-port ${toString cfg.rest.listenPort}"
              ++ lib.optional (cfg.rest.listenAddress != "") "--rest-listen-address ${cfg.rest.listenAddress}"
```

SuperSandro2000 commented on nixos/modules/services/home-automation/openthread-border-router.nix:
```diff
@@ -0,0 +1,216 @@
+    };
+
+    # OTBR needs to publish its addresses via avahi
+    services.avahi.enable = true;
+    services.avahi.publish.enable = true;
+    services.avahi.publish.userServices = true;
```
```suggestion
    services.avahi = {
      enable = true;
      publish = {
        enable = true;
        userServices = true;
      };
    };
```

SuperSandro2000 commented on nixos/modules/services/home-automation/openthread-border-router.nix:
```diff
@@ -0,0 +1,216 @@
+
+    boot.kernel.sysctl = {
+      # Make sure we have ipv6 support, and that forwarding is enabled
+      "net.ipv6.conf.all.disable_ipv6" = 0;
```
Please use networking.enableIPv6 instead

SuperSandro2000 commented on nixos/modules/services/home-automation/openthread-border-router.nix:
```diff
@@ -0,0 +1,216 @@
+    services.openthread-border-router.radio.url =
+      "spinel+hdlc+uart://${cfg.radio.device}?"
+      + lib.concatStringsSep "&" (
+        (lib.optional (cfg.radio.baudRate != 115200) "uart-baudrate=${toString cfg.radio.baudRate}")
+        ++ (lib.optional cfg.radio.flowControl "uart-flow-control")
+        ++ (lib.optional (cfg.radio.extraUrlParams != "") cfg.radio.extraUrlParams)
```
```suggestion
        "uart-baudrate=${toString cfg.radio.baudRate}"
        ++ lib.optional cfg.radio.flowControl "uart-flow-control"
        ++ lib.optional (cfg.radio.extraUrlParams != "") cfg.radio.extraUrlParams
```

SuperSandro2000 commented on nixos/modules/services/home-automation/openthread-border-router.nix:
```diff
@@ -0,0 +1,216 @@
+  };
+
+  config = lib.mkIf cfg.enable {
+
```
```suggestion
```

SuperSandro2000 commented on nixos/modules/services/home-automation/openthread-border-router.nix:
```diff
@@ -0,0 +1,216 @@
+      };
+
+      url = lib.mkOption {
+        type = lib.types.str;
```
We should add an example here

SuperSandro2000 commented on nixos/modules/services/home-automation/openthread-border-router.nix:
```diff
@@ -0,0 +1,216 @@
+        description = "Enable hardware flow control. Ignored if url is manually set.";
+      };
+
+      extraUrlParams = lib.mkOption {
```
```suggestion
      urlQueryString = lib.mkOption {
```

SuperSandro2000 commented on nixos/modules/services/home-automation/openthread-border-router.nix:
```diff
@@ -0,0 +1,216 @@
+      };
+
+      listenPort = lib.mkOption {
+        type = lib.types.int;
```
```suggestion
        type = lib.types.port;
```

SuperSandro2000 commented on nixos/modules/services/home-automation/openthread-border-router.nix:
```diff
@@ -0,0 +1,216 @@
+      enable = lib.mkEnableOption "enable the web interface";
+      listenAddress = lib.mkOption {
+        type = lib.types.str;
+        default = "";
```
What means empty string?

mrene replied:
It defaults to `in6addr_any`, I can set the default to `::` to anchor on that behavior

SuperSandro2000 commented on nixos/modules/services/home-automation/openthread-border-router.nix:
```diff
@@ -0,0 +1,216 @@
+      };
+
+      listenPort = lib.mkOption {
+        type = lib.types.int;
```
```suggestion
        type = lib.types.port;
```

SuperSandro2000 commented on nixos/modules/services/home-automation/openthread-border-router.nix:
```diff
@@ -0,0 +1,216 @@
+    rest = {
+      listenAddress = lib.mkOption {
+        type = lib.types.str;
+        default = "";
```
What means empty string?

mrene replied:
It defaults to `in6addr_any`, I can set the default to `::` to anchor on that behavior

SuperSandro2000 commented on nixos/modules/services/home-automation/openthread-border-router.nix:
```diff
@@ -0,0 +1,216 @@
+      type = lib.types.int;
+      default = 3;
+      description = ''
+        The logging level to use:
```
I understand what is meant here but please write a proper sentence

mrene replied:
Do you mean for the enumeration? Let me know if what I've changed is what you had in mind.

SuperSandro2000 commented on nixos/modules/services/home-automation/openthread-border-router.nix:
```diff
@@ -0,0 +1,216 @@
+  cfg = config.services.openthread-border-router;
+in
+{
+  meta.maintainers = with lib.maintainers; [ mrene ];
```
```suggestion
  meta.maintainers = with lib.maintainers; [ mrene ];

```

SuperSandro2000 reviewed (approved):
LGTM

Pandapip1 commented on nixos/modules/services/home-automation/openthread-border-router.nix:
```diff
@@ -0,0 +1,223 @@
+  meta.maintainers = with lib.maintainers; [ mrene ];
+
+  options.services.openthread-border-router = {
+    enable = lib.mkEnableOption "Enable the OpenThread Border Router";
```
```suggestion
    enable = lib.mkEnableOption "the OpenThread Border Router";
```
`lib.mkEnableOption` prefixes the string with `"Enable "`

Pandapip1 commented on nixos/modules/services/home-automation/openthread-border-router.nix:
```diff
@@ -0,0 +1,223 @@
+      description = "The network interface to create for thread packets";
+    };
+
+    logLevel = lib.mkOption {
+      type = lib.types.int;
+      default = 3;
+      description = ''
+        The level to use when logging messages. The levels are defined as follows:
+          EMERG 0
+          ALERT 1
+          CRIT 2
+          ERR 3
+          WARNING 4
+          NOTICE 5
+          INFO 6
+          DEBUG 7
+      '';
+    };
```
Nit: This would ideally be an enum type

mrene replied:
Good idea!

Pandapip1 commented on nixos/modules/services/home-automation/openthread-border-router.nix:
```diff
@@ -0,0 +1,223 @@
+    };
+
+    web = {
+      enable = lib.mkEnableOption "enable the web interface";
```
```suggestion
      enable = lib.mkEnableOption "the web interface";
```
`lib.mkEnableOption` prefixes the string with `"Enable "`

Pandapip1 commented on nixos/modules/services/home-automation/openthread-border-router.nix:
```diff
@@ -0,0 +1,223 @@
+
+    radio = {
+      device = lib.mkOption {
+        type = lib.types.str;
```
```suggestion
        type = lib.types.path;
```

Pandapip1 commented on nixos/modules/services/home-automation/openthread-border-router.nix:
```diff
@@ -0,0 +1,223 @@
+      device = lib.mkOption {
+        type = lib.types.str;
+        default = "/dev/ttyUSB0";
+        description = "The device name of the serial port of the radio device. Ignored if url is manually set.";
```
```suggestion
        description = "The device name of the serial port of the radio device. Ignored if services.openthread-border-router.radio.url is set.";
```

Pandapip1 commented on nixos/modules/services/home-automation/openthread-border-router.nix:
```diff
@@ -0,0 +1,223 @@
+      baudRate = lib.mkOption {
+        type = lib.types.int;
+        default = 115200;
+        description = "The baud rate of the radio device. Ignored if url is manually set.";
```
```suggestion
        description = "The baud rate of the radio device. Ignored if services.openthread-border-router.radio.url is set.";
```

Pandapip1 commented on nixos/modules/services/home-automation/openthread-border-router.nix:
```diff
@@ -0,0 +1,223 @@
+      flowControl = lib.mkOption {
+        type = lib.types.bool;
+        default = false;
+        description = "Enable hardware flow control. Ignored if url is manually set.";
```
```suggestion
        description = "Enable hardware flow control. Ignored if services.openthread-border-router.radio.url is set.";
```

Pandapip1 commented on nixos/modules/services/home-automation/openthread-border-router.nix:
```diff
@@ -0,0 +1,223 @@
+      };
+
+      url = lib.mkOption {
+        type = lib.types.str;
```
```suggestion
        type = lib.types.nullOr lib.types.str;
```

Pandapip1 commented on nixos/modules/services/home-automation/openthread-border-router.nix:
```diff
@@ -0,0 +1,223 @@
+      urlQueryString = lib.mkOption {
+        type = lib.types.str;
+        default = "";
+        description = "Extra URL query string parameters";
```
```suggestion
        description = "Extra URL query string parameters. Ignored if services.openthread-border-router.radio.url is set.";
```

Pandapip1 commented on nixos/modules/services/home-automation/openthread-border-router.nix:
```diff
@@ -0,0 +1,223 @@
+  };
+
+  config = lib.mkIf cfg.enable {
+    services.openthread-border-router.radio.url =
+      "spinel+hdlc+uart://${cfg.radio.device}?"
+      + lib.concatStringsSep "&" (
+        [ "uart-baudrate=${toString cfg.radio.baudRate}" ]
+        ++ lib.optional cfg.radio.flowControl "uart-flow-control"
+        ++ lib.optional (cfg.radio.urlQueryString != "") cfg.radio.urlQueryString
+      );
```
```suggestion
    services.openthread-border-router.radio.url = lib.mkDefault
      "spinel+hdlc+uart://${cfg.radio.device}?"
      + lib.concatStringsSep "&" (
        [ "uart-baudrate=${toString cfg.radio.baudRate}" ]
        ++ lib.optional cfg.radio.flowControl "uart-flow-control"
        ++ lib.optional (cfg.radio.urlQueryString != "") cfg.radio.urlQueryString
      );
```

Currently this will cause a conflict if `services.openthread-border-router.radio.url` is set elsewhere

mrene replied:
Oh good catch. Do you know if there's a place to write eval-level tests for nixos modules?

mrene commented:
Pushed:
- Applied suggestions
- Convert log level to an enum

mrene commented:
Rebased over master to resolve conflicts in changelog.

Pandapip1 reviewed (approved):
Diff LGTM; not tested

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/2060

milas reviewed (approved):
I have been running this on arm64 / NixOS 24.11 for the past month:

```nix
  services.openthread-border-router = {
    enable = true;
    backboneInterface = "end1";
    radio = {
      device = "/dev/serial/by-id/<your-device-id-here>";
      baudRate = 460800;
      flowControl = true;
    };
  };
```

It's been working great and stable.

Pandapip1 reviewed (commented):
How many of those patches have been upstreamed? The "don't install systemd units" patch can be easily modified to look at a variable:

```
if (SYSTEMD_FOUND AND (NOT DEFINED INSTALL_SYSTEMD_UNIT OR INSTALL_SYSTEMD_UNIT))
    pkg_get_variable(OTBR_SYSTEMD_UNIT_DIR systemd systemdsystemunitdir)
endif()
```

and then `lib.cmakeBool "INSTALL_SYSTEMD_UNIT" false` could be added to `cmakeFlags`

The static libs patch can instead be changed to:

```
if (NOT DEFINED Boost_USE_STATIC_LIBS)
    set(Boost_USE_STATIC_LIBS ON)
endif()
```

and then `lib.cmakeBool "Boost_USE_STATIC_LIBS" stdenv.targetPlatform.isStatic` could be added to `cmakeFlags`.

I'm not sure if the firewall script or the home assistant patch could be upstreamed, but any progress in that direction would be useful.

milas commented:
I opened two upstream PRs with your CMake suggestions:
* https://github.com/openthread/ot-br-posix/pull/2652
* https://github.com/openthread/ot-br-posix/pull/2653

The firewall patch is slightly less straightforward as you noted, I need to dig into that a bit more to understand it from the OTBR side before I open an issue upstream, but it seems tractable as well.

The two Home Assistant-derived patches are a bit murkier.

* https://github.com/home-assistant/addons/pull/3396
  * Desire to add channel monitor to help users find least busy channel
  * This build option means OTBR starts the monitor by default, HASS patch removes that auto-start so that it can be toggled manually
  * Desire to not have it always running is from concern over potential interference
  * Some suggestions
    * Leave as-is -> might consider bringing the patch into nixos/nixpgs, as it's trivial/one-liner
    * Disable channel monitor in NixOS builds for now -> `OPENTHREAD_CONFIG_CHANNEL_MONITOR_ENABLE`
    * Open upstream issue re: behavior - adding `OPENTHREAD_CONFIG_CHANNEL_MONITOR_AUTOSTART` or similar would be straightforward

* https://github.com/home-assistant/addons/pull/3083
  * Patch/changeset is ~1.5yo
  * First issue: uninitialized data -> upstreamed @ https://github.com/openthread/ot-br-posix/pull/1892
  * Second issue: delete/reset fabric
     * Original upstream attempt @ https://github.com/openthread/ot-br-posix/pull/1893
        * This is what we're carrying as a patch right now - "delete dataset"
     * Accepted upstream PR @  https://github.com/openthread/ot-br-posix/pull/1908
        * Revised based on upstream feedback from original approach - "Delete node"

Given that, my suggestion would be to drop the Home Assistant addon patch references entirely:
 * See above: channel monitor patch is trivial and could be "inlined" to this repo (or avoided)
 * Dataset reset is custom to Home-Assistant and will never be upstreamed, so I don't think should be in nixpkgs

milas commented on nixos/modules/services/home-automation/openthread-border-router.nix:
```diff
@@ -0,0 +1,234 @@
+    rest = {
+      listenAddress = lib.mkOption {
+        type = lib.types.str;
+        default = "::";
```
It looks like this should default to `127.0.0.1`. (Specifically, a localhost IP, and it appears that this is the convention throughout NixOS/nixpgs.)

The REST API works fine on IPv4 even though Matter comms themselves use IPv6.

(Users can still pass `::1` if they want to listen on IPv6 localhost instead or `::` as currently to listen on all.)

milas commented on nixos/modules/services/home-automation/openthread-border-router.nix:
```diff
@@ -0,0 +1,234 @@
+      enable = lib.mkEnableOption "the web interface";
+      listenAddress = lib.mkOption {
+        type = lib.types.str;
+        default = "::";
```
Similarly, I think the web interface should default to a local IP, i.e. `127.0.0.1`.

milas commented on nixos/modules/services/home-automation/openthread-border-router.nix:
```diff
@@ -0,0 +1,234 @@
+{
+  meta.maintainers = with lib.maintainers; [ mrene ];
+
+  options.services.openthread-border-router = {
```
I think we also need an `openFirewall` option:
```nix
networking.firewall.allowedUDPPorts = [
    49154
];
```

I don't think it makes sense to have firewall options for REST/web.

In a VM/lab environment, the admin should add `allowedTCPPorts` for `web.listenPort` / `rest.listenPort` themselves.

milas replied:
Opened a discussion about listener ports upstream: https://github.com/orgs/openthread/discussions/11093

milas reviewed (commented):
I pulled the latest changes and rebuilt my config on arm64 without any issues.

While reviewing the diff, I realized I'd made a couple local changes around network security that might be worth incorporating.

Pandapip1 commented on pkgs/by-name/op/openthread-border-router/package.nix:
```diff
@@ -0,0 +1,129 @@
+    license = lib.licenses.bsd3;
+    maintainers = with lib.maintainers; [ mrene ];
+    mainProgram = "ot-ctl";
+    platforms = lib.platforms.linux;
```
```suggestion
    platforms = lib.platforms.unix;
```

If upstream claims that it supports POSIX, we should let it run on all POSIX systems, not just linux.

mrene replied:
While I can get it to compile on aarch64-darwin with the latest apple sdks, it doesn't function correctly (it uses a bogus setsockopt argument when binding to the backbone interface, and it depends on avahi or dns-sd for mdns). 

Would it be preferred to be marked as broken on non-linux while defining platforms as `unix`?

Pandapip1 replied:
I would suggest it be marked broken on darwin. I suspect it would likely work on the BSDs.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/thread-border-router-on-nixos/57909/3

colonelpanic8 commented:
I'm trying to use this with `Home Assistant Connect ZBT-1 (previously known as Home Assistant SkyConnect)`, and seeing this:

```
Jan 18 05:52:03 justin-bieber-creek systemd[1]: Started OpenThread Border Router Agent.
Jan 18 05:52:03 justin-bieber-creek otbr-agent[253962]: [NOTE]-AGENT---: Backbone interface: enp1s0
Jan 18 05:52:05 justin-bieber-creek otbr-agent[253962]: otbr-agent[253962]: 50d.03:40:18.395 [W] P-SpinelDrive-: Wait for response timeout
Jan 18 05:52:05 justin-bieber-creek otbr-agent[253962]: 50d.03:40:18.395 [W] P-SpinelDrive-: Wait for response timeout
Jan 18 05:52:07 justin-bieber-creek otbr-agent[253962]: otbr-agent[253962]: 50d.03:40:20.397 [W] P-SpinelDrive-: Wait for response timeout
Jan 18 05:52:07 justin-bieber-creek otbr-agent[253962]: 50d.03:40:20.397 [W] P-SpinelDrive-: Wait for response timeout
Jan 18 05:52:09 justin-bieber-creek otbr-agent[253962]: otbr-agent[253962]: 50d.03:40:22.399 [W] P-SpinelDrive-: Wait for response timeout
Jan 18 05:52:09 justin-bieber-creek otbr-agent[253962]: otbr-agent[253962]: 50d.03:40:22.399 [C] Platform------: Init() at spinel_driver.cpp:82: Failure
Jan 18 05:52:09 justin-bieber-creek otbr-agent[253962]: 50d.03:40:22.399 [W] P-SpinelDrive-: Wait for response timeout
Jan 18 05:52:09 justin-bieber-creek otbr-agent[253962]: 50d.03:40:22.399 [C] Platform------: Init() at spinel_driver.cpp:82: Failure
Jan 18 05:52:11 justin-bieber-creek otbr-agent[253962]: otbr-agent[253962]: 50d.03:40:24.402 [W] P-SpinelDrive-: Wait for response timeout
Jan 18 05:52:11 justin-bieber-creek otbr-agent[253962]: 50d.03:40:24.402 [W] P-SpinelDrive-: Wait for response timeout
Jan 18 05:52:11 justin-bieber-creek systemd[1]: otbr-agent.service: Main process exited, code=exited, status=1/FAILURE
```

I suspect that it could be because I haven't flashed the thread firmware to the device yet, but I'm struggling to find any information about how to do that manually (all the documentation that I've seen says that the home-assistant add-on should take care of this automatically).

ppw1104 commented:
Tested this, and it's working great so far. Only issue is that after a reboot, the Thread device state is `disabled`. I suspect that it's because the systemd service starts before the network is online, and adding a dependency to `network-online.target` seems to have fixed the issue:

```
  systemd.services.otbr-agent = {
    wants = [ "network-online.target" ];
    after = [ "network-online.target" ];
  };
```

PhilippWoelfel commented:
Tested this, and it's working great so far. Only issue is that after a reboot, the Thread device state is `disabled`. I suspect that it's because the systemd service starts before the network is online, and adding a dependency to `network-online.target` seems to have fixed the issue:

```
  systemd.services.otbr-agent = {
    wants = [ "network-online.target" ];
    after = [ "network-online.target" ];
  };
```

Pandapip1 commented:
With both patches upstreamed, it makes sense to switch the patch fetching to use `fetchpatch`

colonelpanic8 commented:
I'm not going to pretend to have a ton of context here, but https://openthread.io/codelabs/openthread-border-router#ping-from-a-linuxmacos-host

seems to suggest that we probably need to set certain sysctl props:

something like:

```
  boot.kernel.sysctl = {
    # For all interfaces (e.g. if you want to accept RA on all):
    "net.ipv6.conf.all.accept_ra" = lib.mkForce "1";
    "net.ipv6.conf.all.accept_ra_rt_info_max_plen" = lib.mkForce "64";

    # Ensure forwarding is off on all interfaces unless needed
    "net.ipv6.conf.all.forwarding" = lib.mkForce "0";
  };
  ````

Pandapip1 commented:
`mkForce` is unnecessary there, it should just be directly set.

billksun commented:
Is this ready to be merged?

colonelpanic8 commented:
> Is this ready to be merged?

I've been merging this into my nixpkgs flake and using it for a few months now. I would say its good to go!

colonelpanic8 commented:
@SuperSandro2000  @mrene Any way we can push this along and get it merged?

jeantil commented:
hello everyone, 

I used this PR on my nixos install.
- otbr-agent does start and seems to be working
  - a wpan0 interface is created 
  - oc-ctl commands return what looks like valid thread network results 
- otbr-web does start 
  - I was able to form a network through the web ui
  - the status page indicates a thread network is joined
  - topology tab in otbr-web doesn't display anything
- home assistant integration (was it in the scope of the PR ?)
  - home assistant sees the existing thread network  
  - the openthread border router integration is not autoconfigured (maybe this is expected and beyond the scope of this PR) 
  - manually adding the openthread border router integration I can add the thread network as preferred network

I have not managed to add a new device to the network but I think this is beyond the scope of the PR.

jeantil commented on nixos/modules/services/home-automation/openthread-border-router.nix:
```diff
@@ -0,0 +1,234 @@
+    boot.kernel.sysctl = {
+      "net.ipv4.conf.all.forwarding" = 1;
+      "net.ipv6.conf.all.forwarding" = 1;
+      "net.ipv6.ip_forward" = 1;
```
```
sudo sysctl net.ipv6.ip_forward
sysctl: cannot stat /proc/sys/net/ipv6/ip_forward: No such file or directory
```
the option is not listed in https://www.kernel.org/doc/html/latest/networking/ip-sysctl.html either 
according to various non-LLM sources this option is only available for ipv4

jeantil commented on nixos/modules/services/home-automation/openthread-border-router.nix:
```diff
@@ -0,0 +1,234 @@
+      "net.ipv6.ip_forward" = 1;
+
+      # Make sure we accept IPv6 router advertisements from the local network interface
+      "net.ipv6.conf.${cfg.backboneInterface}.accept_ra" = 2;
+      "net.ipv6.conf.${cfg.backboneInterface}.accept_ra_rt_info_max_plen" = 64;
```
This one is much trickier : on my system the settings don't get applied upon boot 
```
> sudo sysctl net.ipv6.conf.all.disable_ipv6 net.ipv4.conf.all.forwarding net.ipv6.conf.all.forwarding net.ipv6.conf.all.accept_ra_rt_info_max_plen net.ipv6.conf.all.accept_ra net.ipv6.conf.wlp1s0.accept_ra net.ipv6.conf.wlp1s0.accept_ra_rt_info_max_plen
Place your finger on the fingerprint reader
net.ipv6.conf.all.disable_ipv6 = 0
net.ipv4.conf.all.forwarding = 1
net.ipv6.conf.all.forwarding = 1
net.ipv6.conf.all.accept_ra_rt_info_max_plen = 0
net.ipv6.conf.all.accept_ra = 1
net.ipv6.conf.wlp1s0.accept_ra = 0
net.ipv6.conf.wlp1s0.accept_ra_rt_info_max_plen = 64
```
however if I reload the settings
```
> sudo sysctl --system
# [...]
net.ipv4.conf.all.forwarding = 1
# [...]
net.ipv6.conf.all.disable_ipv6 = 0
net.ipv6.conf.all.forwarding = 1
net.ipv6.conf.default.disable_ipv6 = 0
net.ipv6.conf.default.use_tempaddr = 2
net.ipv6.conf.wlp1s0.accept_ra = 2
net.ipv6.conf.wlp1s0.accept_ra_rt_info_max_plen = 64
```
then they get applied. I found this thread about a similar problem https://discourse.nixos.org/t/kernel-sysctl-parameter-overriden/57308 
and the diagnostic is correct : on my machine, the network interface is renamed during the boot process 
```
sudo dmesg | grep wlp1s0
[    6.627570] mt7921e 0000:01:00.0 wlp1s0: renamed from wlan0
```
so for the settings to be applied  on boot they would need to target `wlan0`instead of `wlp1s0` however once the system is up and during a nixos switch they need to target `wlp1s0` 

I don't how to work around this apart from adding a config option such as `backboneInterfaceBootName` and changing this config to 
```
     "net.ipv6.conf.${cfg.backboneInterface}.accept_ra" = 2;
      "net.ipv6.conf.${cfg.backboneInterface}.accept_ra_rt_info_max_plen" = 64;
     "net.ipv6.conf.${cfg.backboneInterfaceBootName}.accept_ra" = 2;
      "net.ipv6.conf.${cfg.backboneInterfaceBootName}.accept_ra_rt_info_max_plen" = 64;
``` 
or switching to `.all` which may not be desirable ( I have reached the limit of my ipv6 networking knowledge ):
```
     "net.ipv6.conf.all.accept_ra" = 2;
      "net.ipv6.conf.all.accept_ra_rt_info_max_plen" = 64;
```

jeantil commented on nixos/modules/services/home-automation/openthread-border-router.nix:
```diff
@@ -0,0 +1,234 @@
+
+      listenPort = lib.mkOption {
+        type = lib.types.port;
+        default = 8081;
```
It might be worth noting in the description that running the agent on a different port than the default one will result in a broken topology screen ( the topology screen of the web ui loads the information directly from the agent and the 8081 is currently [hardcoded there](https://github.com/openthread/ot-br-posix/issues/2823) )

mrene committed:
openthread-border-router: unstable-2024-07-26 -> unstable-2025-05-30

TheButlah commented:
is this PR still active / functional?

colonelpanic8 commented:
It looks like it has conflicts now, but yeah it should have just been merged ages ago.

mrene commented:
Pushing a bit of housekeeping:

- Upgrade upstream version to unstable-2025-05-30 (let me know if I should squash this).
- Got rid of the patches since [most of them have been upstreamed](https://github.com/NixOS/nixpkgs/pull/332296#issuecomment-2563813840) or aren't necessary 
- Moved the release notes entry to 25.11
- Removed the bogus ipv6 forward sysctl
- Changed the systemd requires/after to network-online.target instead of dbus.socket (dbus isn't currently activated anyways)

I've been running this version for about a week with no issues so far.

billksun commented on nixos/modules/services/home-automation/openthread-border-router.nix:
```diff
@@ -0,0 +1,233 @@
+
+    package = lib.mkPackageOption pkgs "openthread-border-router" { };
+
+    radioDevice = lib.mkOption {
+      type = lib.types.str;
+      default = "/dev/ttyUSB0";
+      description = "The device name of the serial port of the radio device";
+    };
```
This seems to be a duplicate of `radio.device` and not being used?

mrene replied:
Good catch. Removed.

hatch01 reviewed (approved):
I installed it on my server, and it works like a charm. Thank you very much  @mrene

sanzoghenzo commented on nixos/modules/services/home-automation/openthread-border-router.nix:
```diff
@@ -0,0 +1,227 @@
+    };
+
+    web = {
+      enable = lib.mkEnableOption "the web interface";
```
Is this option used? I cannot see a corresponding mkIf for systemd.services.otbr-web

Pandapip1 commented on nixos/modules/services/home-automation/openthread-border-router.nix:
```diff
@@ -0,0 +1,227 @@
+      };
+
+      # Sync with: src/web/otbr-web.service.in
+      otbr-web = {
```
Yup, this needs a

```suggestion
      otbr-web = lib.mkIf cfg.web.enable {
```

kiesenverseist commented:
I've also been using this on my server with an `slzb-mr` radio, and It's been working great. Since that's a network radio I did have to [reference the home assistant addon](https://github.com/home-assistant/addons/blob/master/openthread_border_router/rootfs/etc/s6-overlay/s6-rc.d/socat-otbr-tcp/run) to see how they handled it. OTBR doesn't seem to be able to directly connect to network URLs. 

I made this small systemd service in my config:
```nix
  systemd.services.otbr-network = {
    description = "otbr network to tty";
    requires = [ "network-online.target" ];
    after = [ "network-online.target" ];
    wantedBy = [ "otbr-agent.service" ];
    before = [ "otbr-agent.service" ];
    path = [pkgs.socat];
    script = ''
      socat -d pty,raw,echo=0,link=/tmp/ttyOTBR,ignoreeof "tcp:192.168.1.35:6638"
    '';
  };
```

I then set my radio URL to `spinel+hdlc+uart:///tmp/ttyOTBR?uart-baudrate=460800`. It'd be nice if this module could handle this usecase too. Maybe a network option under the radio attrset that sets up this service? Then we'd be at feature parity with the home assistant addon. If it's out of scope for this PR, I'd be happy to give it a try myself after this gets merged.

jeantil commented:
> Is it normal behavior to have to set things up with oc-ctl manually?

@sweenu I would say it not.

I was able to form  the thread network or join a thread network and check the network status through otbr's web interface just fine and it reflects the same information as ct-ctl ( check [my config](https://codeberg.org/jeantil/nixos-config/src/branch/main/systems/eidos/services/home-assistant.nix) its in a flake, I define an [otbr overlay](https://codeberg.org/jeantil/nixos-config/src/branch/main/flake.nix#L68) to reference the PR  and [include the module](https://codeberg.org/jeantil/nixos-config/src/branch/main/flake.nix#L86) to make the settings available.) 
I get the topology to display the topology if I use port 8081 for otbr-agent (since that port is hardcoded in the frontend)

All is not perfect yet for my case : I'm only able to provision devices directly from the host that runs otbr using the host's BT adapter and only using a direct manipulation of the matter-server ws as described in  this [post on ha discourse](https://community.home-assistant.io/t/ha-docker-with-otbr-docker/735288/14). 
External provision using a commissioner always fails on network connectivity even though all devices are on the same LAN.  I need to retry provisioning through the HA interface with my latest network settings but that would also be going through the host's BT adapter so not the most convenient.
This is definitely not normal but it can have many causes including my unifi networking gear so you might fare better.

SuperSandro2000 commented on pkgs/by-name/op/openthread-border-router/package.nix:
```diff
@@ -0,0 +1,111 @@
+
+  postInstall = ''
+    mkdir -p $out/bin
+    echo "PWD: "$(pwd)
+    echo "SRC: ${src}"
```
```suggestion
```

hatch01 commented:
Since this issue: #445447
I think you need to add "-DCMAKE_POLICY_VERSION_MINIMUM=3.5" to the package (or maybe update it if it's already fixed upstream).

dotlambda reviewed (commented):
Please add some systemd service hardening.

sweenu commented:
Just wanted to say that I've been using this PR for more than a month with no issues so far.

sweenu commented:
@jeantil 
I have successfully provisioned with my android phone, indeed. You can check my config if you want: https://github.com/sweenu/nixfiles

jeantil commented:
@sweenu thanks for sharing your config and experience !

mrene committed:
openthread-border-router: add systemd service hardening

Rubikoid commented on nixos/modules/services/home-automation/openthread-border-router.nix:
```diff
@@ -0,0 +1,266 @@
+            "AF_INET"
+            "AF_INET6"
+          ];
+          CapabilityBoundingSet = "";
```
There is a problem with these restrictions.

According to https://github.com/openthread/ot-br-posix/blob/main/src/web/web-service/ot_client.cpp#L94, otbr-web connects to daemon over AF_UNIX, to socket in `/run/openthread-wpan0.sock`

mrene replied:
Thanks for testing! I've tested and updated the otbr-web service.

felschr commented on nixos/modules/services/home-automation/openthread-border-router.nix:
```diff
@@ -0,0 +1,266 @@
+        enable = true;
+        userServices = true;
+      };
+    };
```
`avahi` isn't the only services offering multicast DNS, e.g. `systemd-resolved` can also be configured to provide mDNS.

I think the enable options here should at least use `lib.mkDefault` instead to allow users to opt out of avahi, if they use another provider.

felschr commented on nixos/modules/services/home-automation/openthread-border-router.nix:
```diff
@@ -0,0 +1,266 @@
+    package = lib.mkPackageOption pkgs "openthread-border-router" { };
+
+    backboneInterface = lib.mkOption {
+      type = lib.types.str;
```
`--backbone-ifname` [can be specified multiple times](https://github.com/openthread/ot-br-posix/issues/1942#issuecomment-1637942631), so I'd propose changing this to a list.

newAM reviewed (approved):
I tested the package + module end-to-end on `x86_64-linux` with Home Assistant, `services.matter-server`, a home-assistant ZBT-1 dongle, and a Nanoleaf essentials bulb, everything is working great.  Thank you to everyone who contributed to this pull-request!

mrene commented:
Superseeded by #502388
